Create and edit vector layers

HaudinFlorence opened the pull request:
Addresses some of #395 ; resolves #775

Create and edit vector layers


https://github.com/user-attachments/assets/d8708804-06f0-4859-8e7e-13be39665600



- the creation of a new vector layer with empty geojson source is possible using the `+` button of the toolbar, then selecting `add new vector layer` and `new geoJSON Layer` leading to the opening of the dedicated Dialog. 
Changes have been done to enable situations where no path is provided
- 3 geometries are enabled for draw interactions : `Point`, `LineString` and `Polygon`



<!-- readthedocs-preview jupytergis start -->
---
📚 Documentation preview: https://jupytergis--692.org.readthedocs.build/en/692/
💡 JupyterLite preview: https://jupytergis--692.org.readthedocs.build/en/692/lite

<!-- readthedocs-preview jupytergis end -->

HaudinFlorence committed:
Add a toggle button to enable to add a draw vector layer.
Add a isDrawVectorLayerEnabled property and a drawVectorLayerChanged signal to the JupyterGISModel.
Add a method _handleVectorLayerChange to update the state of isDrawVectorLayerEnabled.
Add VectorLayerDropdown react component to be used in an overlay when the draw vector layer button is toggled.

mfisher87 commented:
Yess! This feature is very in-demand :)  :clap:

HaudinFlorence commented:
> Yess! This feature is very in-demand :) 👏
Thanks for having added the link to the related issue @mfisher87

martinRenou commented on packages/base/src/constants.ts:
```diff
@@ -40,6 +40,17 @@ export namespace CommandIDs {
   export const buffer = 'jupytergis:buffer';
   export const dissolve = 'jupytergis:dissolve';
 
+  // Layers only commands
+  export const newRasterLayer = 'jupytergis:newRasterLayer';
+  export const newVectorLayer = 'jupytergis:newVectorLayer';
+  export const newHillshadeLayer = 'jupytergis:newHillshadeLayer';
+  export const newImageLayer = 'jupytergis:newImageLayer';
+  export const newVideoLayer = 'jupytergis:newVideoLayer';
+  export const newShapefileLayer = 'jupytergis:newShapefileLayer';
+  export const newWebGlTileLayer = 'jupytergis:newWebGlTileLayer';
+  export const newHeatmapLayer = 'jupytergis:newHeatmapLayer';
+  export const newDrawVectorLayer = 'jupytergis:newDrawVectorLayer';
+
```
Probably some rebase leftover

```suggestion
```

HaudinFlorence replied:
Yes. I just realized before your comment that I had some rebase related issues.

HaudinFlorence committed:
Remove the VectorLayerDropdown component and move the logics that was inside it to the MainView class.

HaudinFlorence committed:
Change _handleDrawVectorLayerChanged method to _updateIsDrawVectorLayerEnabled.

HaudinFlorence committed:
Add logics to create the vector layer and to add it to the map.

HaudinFlorence committed:
Add logics to remove the draw interaction when isDrawVectorLayerEnabled is false.

HaudinFlorence committed:
Add an enabled method to the newVDrawVectorLayer, update the toggle method and add the command to the _notifyCommands of the LeftPanelWidget class.

HaudinFlorence committed:
Add logics to use the source of the currently selected layer in the _handleDrawGeometryTypeChange method of the MainView class.

HaudinFlorence committed:
Update the _handleDrawGeometryTypeChange method with removing the creation of a new instance of vectorLayer and the addition of this layer to the map.
Add style to the draw interaction.

martinRenou commented on packages/base/src/commands.ts:
```diff
@@ -878,6 +878,56 @@ export function addCommands(
+        return false;
+      }
+
+      const canDrawVectorLayer = ['VectorLayer'].includes(selectedLayer.type);
```
For the first version, let's allow only editing vector layers with a `data` property (embedded data in the jgis file).

Saving edits of files backed by external url or local path requires more discussion (for those files, we could save the diff, and allow exporting the result to a new file on the server).

```suggestion
      const canDrawVectorLayer = selectedSource.type === 'GeoJSONSource' && selectedSource.data;
```

martinRenou commented on packages/schema/src/interfaces.ts:
```diff
@@ -232,6 +233,9 @@ export interface IJupyterGISModel extends DocumentRegistry.IModel {
   disposed: ISignal<any, void>;
+  isDrawVectorLayerEnabled: boolean;
+  updateIsDrawVectorLayerEnabled(): void;
+  selectedVectorLayerSourceId: string;
```
Introducing this would require us to make sure that value is always up-to-date, which is unneeded added complexity.

I don't think we need to introduce this, we already have a way to get the selected layer (then its source) from the main view:

```typescript
const localState = this._model?.sharedModel.awareness.getLocalState();
const selectedLayer = localState?.selected?.value;
```

HaudinFlorence replied:
@martinRenou Thanks for the comment. I will change this.

martinRenou commented on test-results/.last-run.json:
```diff
@@ -0,0 +1,4 @@
+{
```
This file should be removed (maybe the full dir should be git-ignored)

martinRenou commented on packages/base/src/commands.ts:
```diff
@@ -878,6 +878,56 @@ export function addCommands(
+    label: trans.__('Create New Draw Vector Layer'),
+    isToggled: () => {
+      const selectedLayer = getSingleSelectedLayer(tracker);
+      if (!selectedLayer) {
```
In case there is no selected layer, what would make the most sense in terms of UX would be to automatically create a new GeoJSONSource (with an associated layer) and start editing that layer.

martinRenou reviewed (changes requested):
Thanks! I made some suggestions

HaudinFlorence committed:
Update packages/base/src/commands.ts

Co-authored-by: martinRenou <martin.renou@gmail.com>

HaudinFlorence committed:
Update newDrawVectorLayer command.

HaudinFlorence committed:
Remove test-results/.last-run.json from the git history tracking and add test-results to the .gitignore file.

HaudinFlorence committed:
Update the newDrawVectorLayer command and the _handleDrawGeometryTypeChange method to update the source in the JupyterGISDoc.

HaudinFlorence committed:
Remove the enabled method in the the newDrawVectorLayer command.
Create a new empty layer with a GeoJSONSource when there is no selected layer.
Add some logics to ensure that the overlay is properly removed when the selected layer is switched and doesn't have the right source type and some embedded data.

HaudinFlorence committed:
Update _handleDrawGeometryTypeChange with replacing the drawend event by a on change event on the layer source. Use a GeoJSON instance and its writeFeatureObject method for the features attribute of the updatedData.

HaudinFlorence committed:
Change the ui : use the toolbar button for the creation of a new empty vector layer and add a context menu for an already existing vector layer.
Update _handleDrawGeometryTypeChange method of MainView class to properly deal with the case where there is a selected vector layer and the case where there isn't.
Add a new command addNewDrawFeatures enabled in the context menu of a Vector Layer.
Move the logics for the creation of an empty layer with an empty geojson source into a dedicated method createEmptyVectorLayerWithGeoJSONSource in the JupyterGISModel class.

HaudinFlorence committed:
Update the GeoJSONSourcePropertiesForm class to enable the possibility to not have a path provided and to enable the creation of an empty layer with geoJSON source with embedded data.

HaudinFlorence committed:
Change the ui to create a new empty vector layer by clicking on the plus toolbar button and using the already existing dialog and form.
  Modify the logics to enable situations where no path is provided in the dialog.
  Remove the command addNewDrawVectorLayer and the toolbar button.
  Remove method createEmptyVectorLayerWithGeoJSONSource from JupyterGISModel class and IJupyterGISModel interface.
  Update _handleDrawGeometryTypeChange method from the MainView class.
  Add a condition if GeoJSONSourcePropertiesForm for the case where there is no path.

HaudinFlorence committed:
Provide a onFormSubmit method to the GeoJSONSourcePropertiesForm and remove the condition for no path case that was wrongly in the onFormSubmit method of the parent class PathBasedSourceProperties.
Update _handleDrawGeometryTypeChange in MainView class.

HaudinFlorence committed:
Try to modify the path description in the processSchema method of GeoJSONSourcePropertiesForm.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

martinRenou reviewed (changes requested):
Thanks! Works great!

I left some suggestions.

Another suggestion I could make:
- set `isDrawVectorLayerEnabled` when the layer on which we're drawing is unselected, to match the behavior of e.g. the identify feature

HaudinFlorence committed:
Apply suggestions from code review

Co-authored-by: martinRenou <martin.renou@gmail.com>

HaudinFlorence commented on packages/base/src/constants.ts:
```diff
@@ -41,6 +41,16 @@ export namespace CommandIDs {
+  export const newShapefileLayer = 'jupytergis:newShapefileLayer';
+  export const newWebGlTileLayer = 'jupytergis:newWebGlTileLayer';
+  export const newHeatmapLayer = 'jupytergis:newHeatmapLayer';
+
```
@martinRenou. Yes this is a left over from a rebase. Sorry for this.

HaudinFlorence committed:
Take review comments into account.

HaudinFlorence committed:
Restore toggle toolbar button for the toggleDrawFeatures command.
In leftPanel,use notifyCommandChanged for all commands.
Fix bug regarding the overlay display when a draw vector layer is removed.
Enable the toggleDrawFeature command only when the layer is a draw vector layer one.

HaudinFlorence committed:
Refactor the code.

HaudinFlorence committed:
Update methods related to draw interactions in MainView class.

HaudinFlorence committed:
Update all interactions in a single updateInteractions.
Add Select interaction.
Set activity of the different interactions properly.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

HaudinFlorence committed:
Apply suggestions from code review

Co-authored-by: martinRenou <martin.renou@gmail.com>

HaudinFlorence commented on packages/base/src/mainview/mainView.tsx:
```diff
@@ -2009,6 +2093,183 @@ export class MainView extends React.Component<IProps, IStates> {
+    this._Map.removeInteraction(this._modify);
+  };
+
+  private _onVectorSourceChangeHandler = () => {
+    this._onVectorSourceChange();
+  };
```
I will remove it.

HaudinFlorence committed:
Take review comments into account.

HaudinFlorence committed:
Add a check on the type of the layer in onSelectedLayerChange to prevent calling editVectorLayer on irrelevant layers.

martinRenou commented:
Thanks!

@mfisher87 would you have time to have a look?

How about we tackle the `VectorLayer` type property separately from this PR?

mfisher87 commented:
>How about we tackle the `VectorLayer` type property separately from this PR?

That sounds good to me!
